triton-constraint-builder/src/codegen.rs

```diff
@@ -743,13 +745,36 @@ impl TasmBackend {
             return self.load_node(constraint);
         };
 
+        // Use correct operator as lhs/rhs can be either a base-field element
+        // or an extension-field element. Since the TASM code evaluates
+        // out-of-domain rows, all inputs are extension field elements. So the
+        // only expression that is a base-field element is a base-field
+        // constant.
+        let lhs_bfe_const = match lhs.borrow().expression {
+            CircuitExpression::BConst(bfe) => Some(bfe),
+            _ => None,
+        };
+        let rhs_bfe_const = match rhs.borrow().expression {
+            CircuitExpression::BConst(bfe) => Some(bfe),
+            _ => None,
+        };
         let lhs = self.evaluate_single_node(&lhs.borrow());
         let rhs = self.evaluate_single_node(&rhs.borrow());
-        let binop = match binop {
-            BinOp::Add => instr!(XxAdd),
-            BinOp::Mul => instr!(XxMul),
-        };
-        [lhs, rhs, binop].concat()
+
+        match (binop, lhs_bfe_const, rhs_bfe_const) {
+            (_, Some(_), Some(_)) => unreachable!("Constant folding should have handled this"),
+            (binop, None, None) => {
+                let binop = match binop {
+                    BinOp::Add => instr!(XxAdd),
+                    BinOp::Mul => instr!(XxMul),
+                };
+                [lhs, rhs, binop].concat()
+            }
+            (BinOp::Add, None, Some(bfe)) => [lhs, instr!(AddI(bfe))].concat(),
+            (BinOp::Add, Some(bfe), None) => [rhs, instr!(AddI(bfe))].concat(),
+            (BinOp::Mul, None, Some(_)) => [lhs, rhs, instr!(XbMul)].concat(),
+            (BinOp::Mul, Some(_), None) => [rhs, lhs, instr!(XbMul)].concat(),
+        }
```

Comment:
I was confused by both comment and code. I suggest the following:[^diff]

```suggestion
        let tokenized_lhs = self.evaluate_single_node(&lhs.borrow());
        let tokenized_rhs = self.evaluate_single_node(&rhs.borrow());
        let tokenized_binop = match binop {
            BinOp::Add => instr!(XxAdd),
            BinOp::Mul => instr!(XxMul),
        };

        // Use more efficient instructions if one side is a base field element.
        // Applying domain-specific knowledge, `CircuitExpression::Input`s can never be
        // base field elements: the constraints are evaluated on out-of-domain rows.
        let extract_bfe_const = |circuit| match circuit.borrow().expression {
            CircuitExpression::BConst(bfe) => Some(bfe),
            _ => None,
        };

        match (binop, extract_bfe_const(lhs), extract_bfe_const(rhs)) {
            (_, Some(_), Some(_)) => panic!("Constant folding should have handled this"),
            (_, None, None) => [tokenized_lhs, tokenized_rhs, tokenized_binop].concat(),
            (BinOp::Add, None, Some(c)) => [tokenized_lhs, instr!(AddI(c))].concat(),
            (BinOp::Add, Some(c), None) => [tokenized_rhs, instr!(AddI(c))].concat(),
            (BinOp::Mul, None, Some(_)) => [tokenized_lhs, tokenized_rhs, instr!(XbMul)].concat(),
            (BinOp::Mul, Some(_), None) => [tokenized_rhs, tokenized_lhs, instr!(XbMul)].concat(),
        }
```

[^diff]: Since the diff includes deleted lines, the interface does not allow me to make a suggestion for the range of lines I suggest to replace. The range starts where the current comment starts and ends as indicated by github's interface.